aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -102,15 +127,24 @@ struct DispatchKeyExtractor final {
     return impl::dispatchTypeId(typeSet);
   }
 
-  explicit DispatchKeyExtractor(size_t num_args)
-  : num_args_(num_args) {}
+  explicit DispatchKeyExtractor(
+    size_t num_args,
+    c10::optional<size_t> scalar_type_index,
+    c10::optional<size_t> layout_index)
+  : num_args_(num_args),
+    scalar_type_index_(scalar_type_index),
+    layout_index_(layout_index) {}
 
   // this is caching the index so we don't have to parse the schema inputs
   // again and again for each dispatcher lookup.
   // num_args_ is allowed to be zero; that just means you must do the
   // fallthrough
   // TODO: a potential optimization is to store a bitfield of arg locations,
   size_t num_args_;
+  // this is caching the index for ScalarType and LayoutType for args inside
+  // function schema.
+  c10::optional<size_t> scalar_type_index_;
```

Comment:
We only want to consider ScalarType for dispatch if all TensorOptions arguments are present and in the correct order, right? So only then, this optional should be set. And that also means that you don't need to store `layout_index_`, because you can calculate it by offset from scalar_type_index_.

Reply:
yes. I forget that. I will just use one index here.